File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
def solution(board):
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

    robot_pos = None
    target_pos = None
    for i in range(len(board)):
        for j in range(len(board[0])):
            if board[i][j] == 'R': robot_pos = (i, j)
            elif board[i][j] == 'G': target_pos = (i, j)

    queue = [(robot_pos, 0)]
    visited = set()
    visited.add(robot_pos)
    while queue:
        curr_pos, move_count = queue.pop(0)
        if curr_pos == target_pos: return move_count
        for dx, dy in directions:
            x, y = curr_pos
            while 0 <= x+dx < len(board) and 0 <= y+dy < len(board[0]) and board[x+dx][y+dy] != 'D':
                x += dx
                y += dy
            if (x, y) not in visited:
                visited.add((x, y))
                queue.append(((x, y), move_count + 1))

    return -1 

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.js
const direction = [
  [0, 1],
  [1, 0],
  [0, -1],
  [-1, 0],
];

let n, m;
function solution(board) {
  n = board.length;
  m = board[0].length;
  let sx, sy, ex, ey;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (board[i][j] === "R") {
        sx = i;
        sy = j;
      }
      if (board[i][j] === "G") {
        ex = i;
        ey = j;
      }
    }
  }

  const discovered = Array.from(Array(n), () => Array(m).fill(false));
  discovered[sx][sy] = true;
  let queue = [[sx, sy, 0]];
  while (queue.length !== 0) {
    const size = queue.length;
    let nextQueue = [];

    for (let i = 0; i < size; i++) {
      const [x, y, cnt] = queue.pop();

      for (const [dx, dy] of direction) {
        let nx = x + dx;
        let ny = y + dy;
        while (check(nx, ny) && board[nx][ny] !== "D") {
          nx += dx;
          ny += dy;
        }
        nx -= dx;
        ny -= dy;
        if (board[nx][ny] === "G") {
          return cnt + 1;
        }
        if (!discovered[nx][ny]) {
          discovered[nx][ny] = true;
          nextQueue.push([nx, ny, cnt + 1]);
        }
      }
    }
    queue = nextQueue;
  }
  return -1;
}

const check = (nx, ny) => {
  return 0 <= nx && nx < n && 0 <= ny && ny < m;
};